File: backend/app/clients/searchgov.py
import logging

from .base import BaseAPIClient
from ..config import get_settings
from ..models.schemas import SourceItem

logger = logging.getLogger(__name__)


class SearchGovClient(BaseAPIClient):
    def __init__(self):
        settings = get_settings()
        super().__init__(base_url=settings.searchgov_base_url)
        self.affiliate = settings.searchgov_affiliate
        self.access_key = settings.searchgov_access_key

    @property
    def is_configured(self) -> bool:
        return bool(self.affiliate and self.access_key)

    def _normalize_result(self, result: dict) -> SourceItem:
        return SourceItem(
            source_type="searchgov",
            id=result.get("link", "")[:100],
            title=result.get("title", "Untitled"),
            agency=None,
            date=result.get("publication_date") or result.get("created_at"),
            url=result.get("link", ""),
            excerpt=result.get("snippet"),
            content_type="web_result",
            raw=result,
        )

    async def search(
        self,
        query: str,
        enable_highlighting: bool = False,
        limit: int = 10,
        offset: int = 0,
    ) -> tuple[list[dict], list[SourceItem]]:
        params = {
            "affiliate": self.affiliate,
            "access_key": self.access_key,
            "query": query,
            "enable_highlighting": str(enable_highlighting).lower(),
            "limit": limit,
            "offset": offset,
        }

        url = f"{self.base_url}/results/i14y"
        logger.info(f"Searching Search.gov: {query}")

        data = await self._request_with_retry(
            method="GET",
            url=url,
            params=params,
        )

        web_results = data.get("web", {}).get("results", [])
        sources = [self._normalize_result(r) for r in web_results]

        return web_results, sources

File: backend-ts/src/clients/searchgov.ts
import { BaseAPIClient } from "./base.js";
import { getSettings } from "../config.js";
import type { SourceItem } from "../models/schemas.js";

export class SearchGovClient extends BaseAPIClient {
    private affiliate: string;
    private accessKey: string;

    constructor() {
        const settings = getSettings();
        super(settings.searchGovBaseUrl);
        this.affiliate = settings.searchGovAffiliate;
        this.accessKey = settings.searchGovAccessKey;
    }

    get isConfigured(): boolean {
        return Boolean(this.affiliate && this.accessKey);
    }

    private normalizeResult(result: Record<string, unknown>): SourceItem {
        return {
            source_type: "searchgov",
            id: ((result.link as string) || "").slice(0, 100),
            title: (result.title as string) || "Untitled",
            agency: null,
            date:
                (result.publication_date as string) ||
                (result.created_at as string) ||
                null,
            url: (result.link as string) || "",
            excerpt: result.snippet as string || null,
            content_type: "web_result",
            raw: result,
        };
    }

    async search(options: {
        query: string;
        enableHighlighting?: boolean;
        limit?: number;
        offset?: number;
    }): Promise<{ results: Record<string, unknown>[]; sources: SourceItem[] }> {
        const { query, enableHighlighting = false, limit = 10, offset = 0 } = options;

        const params: Record<string, string | number | boolean> = {
            affiliate: this.affiliate,
            access_key: this.accessKey,
            query,
            enable_highlighting: String(enableHighlighting).toLowerCase(),
            limit,
            offset,
        };

        const url = `${this.baseUrl}/results/i14y`;

        const data = await this.requestWithRetry<{
            web?: { results?: Record<string, unknown>[] };
        }>({
            url,
            params,
        });

        const webResults = data.web?.results || [];
        const sources = webResults.map((r) => this.normalizeResult(r));

        return { results: webResults, sources };
    }
}
